### ISSUE TITLE
Little questions

### ISSUE THREAD

**Issue by elyograg** [remkop/picocli#2021]
I am working on rewriting the startup script for Apache Solr in Java.  I've encountered some things in the automatic usage generation I can't figure out.  I was going to post this in the Google group, but it recommended coming here instead.

* Some of the options are indented oddly.  Is there any way to have all the options at the same indentation level?  I think what's making the odd indentation is that some of the options have two characters instead of one.
* Is it possible to override the usage text for a particular option?  The way it represents the -D option is not how I intend it to work.  I would like it to say `-Dprop="custom value"` as I have in the description and show the `-D` part of that in the brown color.
* The "ERROR StatusLogger" messages in the screenshot are new, and I do not know what changed to cause them to be logged.  Does anyone know how to eliminate them?
* It shows `-option=value` for all options, but `-option value` is the way I actually intend to use this, and that does work. Can that be adjusted so it either shows the space option only or shows both?
 
![image](https://user-images.githubusercontent.com/4984827/236498631-2ff92503-b2fc-4f25-99cc-a21b0b662a92.png)

**Comment by remkop** [comment]
Hi @elyograg Thank you for raising this!

### Indentation
The indentation: your deduction is correct, it has to do with the option name. Picocli gives special treatment to [POSIX-style options](https://picocli.info/#_short_posix_options), which are one-character options, preceded by a dash or hyphen. POSIX-style options can be concatenated into a single "word" on the command line, like the `-v`, `-x` and `-f` options of the `tar` command can be combined like this: `tar -xvfSomeFile.tar` (equivalent to `tar -x -v -f SomeFile.tar`).
This only works for one-character options though.

Many tools have adopted the convention to give "common" options both a short and a long name, and give "uncommon" options only a long name, and the indentation can serve as a visual guideline to new users as to which options they may be most interested in.
It is possible to customize this, but it will require some code. You can find an example [here](https://github.com/remkop/picocli/blob/main/picocli-examples/src/main/java/picocli/examples/leftalign/LeftAlignOptions.java).

Alternatively you can consider adding short (POSIX) option names to these options, like `-g`, `-z`, etc.

### Option label
By default, picocli shows `-D<properties>` because the name of the field is `properties`. You can override this by setting a [paramLabel](https://picocli.info/#_parameter_labels). Something like this:

```java
@Option(name = "-D", paramLabel="prop=\"custom value\"")
Properties properties;
```

### Errors about conversion specifier
(NOTE TO SELF) I need to improve the [documentation for this](https://picocli.info/#_format_specifiers). Specifically, I need to add that error text so that people can search the manual for that error and find it.

Long story short: I believe your usage help descriptions have one or more single `%` characters in there somewhere.
Please escape it with another `%`, so your descriptions look like this `xxxx %% xxxx`. That should make the errors go away.

### Custom option-parameter separators
The parser actually accepts both but displays the `=`.

The easiest way is described here: https://picocli.info/#_option_parameter_separators
but that also impacts parsing so that may not do what you want.

There may be something in the [Examples](https://github.com/remkop/picocli/tree/main/picocli-examples/src/main/java/picocli/examples). I have to run now and will dig some more later.

**Comment by elyograg** [comment]
The separator option got me something that looks a LOT better.  See screenshot.

At first I thought I wasn't going to be able to use the left-align code, because the subcommand class does not have a main method.  I added it to the class with my main method, which is also a @Command, and the setting DID get applied to the subcommands too!  Thanks!

I am slightly annoyed that it put two spaces between -D and the example option when I changed the separator, but that's VERY minor.  At least I don't have to manually update usage text when something changes!

It wrapped the first Usage section in the middle of `[sp <stopPort>]`as well as the middle of the -D text ... is there an line wrap option for keeping that text together?

![image](https://user-images.githubusercontent.com/4984827/236583894-29ff6df3-558e-4ae8-b929-280ee0333010.png)

**Comment by remkop** [comment]
About the wrapping at 80 characters width:
You can tell picocli that the help should use up the [full terminal width](https://picocli.info/#_auto_terminal_width) with `@Command(usageHelpAutoWidth = true)`.
You can also [set the width manually](https://picocli.info/#_usage_width) with `@Command(usageHelpWidth = NUMBER)` in the annotations (and you can do both, so the new manual width is used instead of 80 characters when picocli cannot detect the terminal width).

I am glad you like how the usage help message looks after setting the separator option to `" "` space.
Note that this unfortunately also means that the parser will no longer accept input like `--listen-port=8984` (please try this, I believe the parser now only accepts a space between the option name and the option parameter), which may be an unpleasant surprise to your users.
It is not nice that you have to make this trade-off.
If you feel like contributing support for a _display-only_ separator, pull requests are very welcome! 😅 
(My time to work on picocli is extremely limited these days.)

**Comment by elyograg** [comment]
The errors turned out to be caused by log4j2, not picocli.  I use slf4j with the log4j2 binding.  It turns out that if the log4j-1.2-api jar is present, the program outputs those errors.  I sure hope I don't need a dependency that is explicitly bound to log4j 1.2, because sf4j's log4j-over-slf4j jar causes the errors too.

**Comment by remkop** [comment]
Thanks for that, I just realized that I had put some special logic in place to show this much nicer message:  

> "Using raw String: '%n' format strings have not been replaced with newlines. " +
> "Please ensure to escape '%' characters with another '%'."));

I had forgotten about that. 😅

**Comment by elyograg** [comment]
> About the wrapping at 80 characters width: You can tell picocli that the help should use up the [full terminal width](https://picocli.info/#_auto_terminal_width) with `@Command(usageHelpAutoWidth = true)`. You can also [set the width manually](https://picocli.info/#_usage_width) with `@Command(usageHelpWidth = NUMBER)` in the annotations (and you can do both, so the new manual width is used instead of 80 characters when picocli cannot detect the terminal width).

Exactly what column it wraps at isn't too important to me.  I just want it to keep the option and its value together on the same line -- treat it all as a single word.  So those lines would look like this instead:

```
Usage: solr start [-cf] [-gc <gc>] [-m <heapSize>] [-p <listenPort>]
                  [-sp <stopPort>] [-z <zkHost>] [-zp <zkPort>]
                  [-D prop="prop value"]...
```

> I am glad you like how the usage help message looks after setting the separator option to `" "` space. Note that this unfortunately also means that the parser will no longer accept input like `--listen-port=8984` (please try this, I believe the parser now only accepts a space between the option name and the option parameter), which may be an unpleasant surprise to your users. It is not nice that you have to make this trade-off. If you feel like contributing support for a _display-only_ separator, pull requests are very welcome! sweat_smile (My time to work on picocli is extremely limited these days.)

I will be fine with only a space working ... I doubt the current parsing in bash works with equal signs.  It would be very cool to have that only affect display, but that's not critical.  If I ever find enough time, I can take a look at your code to see if I can do anything.

**Comment by remkop** [comment]
> > About the wrapping at 80 characters width: You can tell picocli that the help should use up the [full terminal width](https://picocli.info/#_auto_terminal_width) with `@Command(usageHelpAutoWidth = true)`. You can also [set the width manually](https://picocli.info/#_usage_width) with `@Command(usageHelpWidth = NUMBER)` in the annotations (and you can do both, so the new manual width is used instead of 80 characters when picocli cannot detect the terminal width).
> 
> Exactly what column it wraps at isn't too important to me. I just want it to keep the option and its value together on the same line -- treat it all as a single word. So those lines would look like this instead:
> 
> ```
> Usage: solr start [-cf] [-gc <gc>] [-m <heapSize>] [-p <listenPort>]
>                   [-sp <stopPort>] [-z <zkHost>] [-zp <zkPort>]
>                   [-D prop="prop value"]...
> ```

Oh I see, you are talking about the synopsis.
No, that is unfortunately a side effect of removing the "=" character, they are now separate words so the break iterator will split them if they don't fit on a single line. The best thing may be is to use the full width of the terminal so the lines are longer there is less wrapping going on.

> > I am glad you like how the usage help message looks after setting the separator option to `" "` space. Note that this unfortunately also means that the parser will no longer accept input like `--listen-port=8984` (please try this, I believe the parser now only accepts a space between the option name and the option parameter), which may be an unpleasant surprise to your users. It is not nice that you have to make this trade-off. If you feel like contributing support for a _display-only_ separator, pull requests are very welcome! sweat_smile (My time to work on picocli is extremely limited these days.)
> 
> I will be fine with only a space working ... I doubt the current parsing in bash works with equal signs. It would be very cool to have that only affect display, but that's not critical. If I ever find enough time, I can take a look at your code to see if I can do anything.

Current parsing should work with equal signs in any shell; if not please let me know, that is a bug. 

I believe most end users would expect `--some-option=value` to work but if you are okay with that no longer working, that is fine of course.

**Comment by elyograg** [comment]
> > I will be fine with only a space working ... I doubt the current parsing in bash works with equal signs. It would be very cool to have that only affect display, but that's not critical. If I ever find enough time, I can take a look at your code to see if I can do anything.
> 
> Current parsing should work with equal signs in any shell; if not please let me know, that is a bug.

The script I am trying to convert to Java is written in bash, and I am pretty sure it is not able to do equal signs.  From what I have seen, it uses $1, $2, and shift.  So I am not worried about equal signs not working. For the start command, I can already offer more option functionality than the existing script has.  I am really liking this project!

**Comment by remkop** [comment]
Thank you, that’s great to hear!

**Comment by elyograg** [comment]
I decided to not fight the equal sign.  Created some footer text to cover other possibilities.
![image](https://user-images.githubusercontent.com/4984827/236698105-6eb7c73e-a102-4bf7-aa09-2ca7b3cabd19.png)
